search/features_layer_matcher.hpp

```diff
@@ -181,36 +185,79 @@ class FeaturesLayerMatcher
   {
+    BailIfCancelled(m_cancellable);
+
     ASSERT_EQUAL(child.m_type, Model::TYPE_POI, ());
     ASSERT_EQUAL(parent.m_type, Model::TYPE_STREET, ());
 
     auto const & pois = *child.m_sortedFeatures;
     auto const & streets = *parent.m_sortedFeatures;
 
-    // When the number of POIs is less than the number of STREETs,
-    // it's faster to check nearby streets for POIs.
-    if (pois.size() < streets.size())
+    vector<PointRectMatcher::PointIdPair> poiCenters;
+    poiCenters.reserve(pois.size());
+
+    for (size_t i = 0; i < pois.size(); ++i)
+    {
+      FeatureType poiFt;
+      GetByIndex(pois[i], poiFt);
+      poiCenters.emplace_back(feature::GetCenter(poiFt, FeatureType::WORST_GEOMETRY), i /* id */);
+    }
+
+    vector<PointRectMatcher::RectIdPair> streetRects;
+    streetRects.reserve(streets.size());
+
+    vector<ProjectionOnStreetCalculator> streetProjectors;
+    streetProjectors.reserve(streets.size());
+
+    for (size_t i = 0; i < streets.size(); ++i)
     {
-      for (uint32_t poiId : pois)
+      FeatureType streetFt;
+      GetByIndex(streets[i], streetFt);
+
+      streetFt.ParseGeometry(FeatureType::WORST_GEOMETRY);
+
+      m2::RectD inflationRect;
+      // Any point is good enough here, and feature::GetCenter would re-read the geometry.
+      if (streetFt.GetPointsCount() > 0)
       {
-        for (auto const & street : GetNearbyStreets(poiId))
-        {
-          if (street.m_distanceMeters > kStreetRadiusMeters)
-            break;
+        inflationRect = MercatorBounds::RectByCenterXYAndSizeInMeters(streetFt.GetPoint(0),
+                                                                      0.5 * kStreetRadiusMeters);
+      }
 
-          uint32_t const streetId = street.m_id.m_index;
-          if (binary_search(streets.begin(), streets.end(), streetId))
-            fn(poiId, streetId);
-        }
+      for (size_t j = 0; j + 1 < streetFt.GetPointsCount(); ++j)
+      {
+        auto const & p1 = streetFt.GetPoint(j);
+        auto const & p2 = streetFt.GetPoint(j + 1);
+        m2::RectD rect(p1, p2);
+        rect.Inflate(inflationRect.SizeX(), inflationRect.SizeY());
+        streetRects.emplace_back(rect, i /* id */);
       }
-      return;
-    }
 
-    for (uint32_t streetId : streets)
-    {
-      BailIfCancelled(m_cancellable);
-      m_loader.ForEachInVicinity(streetId, pois, kStreetRadiusMeters, bind(fn, _1, streetId));
+      vector<m2::PointD> streetPoints;
+      streetPoints.reserve(streetFt.GetPointsCount());
+      for (size_t j = 0; j < streetFt.GetPointsCount(); ++j)
+        streetPoints.emplace_back(streetFt.GetPoint(j));
+      streetProjectors.emplace_back(streetPoints);
     }
+
+    BailIfCancelled(m_cancellable);
+    // The matcher may call |fn| more than once if a POI falls inside
+    // the bounding rectangles of several street segments.
+    set<pair<size_t, size_t>> fnAlreadyCalled;
```

Comment:
Please, replace set of pairs by a vector of bools, the size of vector should be the same as the number of pois.

Reply:
I move the check up the call tree.